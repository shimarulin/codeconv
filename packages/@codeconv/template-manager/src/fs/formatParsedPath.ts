import { format, join } from 'node:path'
import { FileMeta } from './readFsThree.js'

export function formatParsedPath (parsedPath: FileMeta): string {
  return join(parsedPath.root, format(parsedPath))
}
