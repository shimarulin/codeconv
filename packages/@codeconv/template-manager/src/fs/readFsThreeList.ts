import { reduce } from 'ramda'
import { type FileMeta, readFsThree } from './readFsThree.js'

const mergeFileLists = reduce<FileMeta[], FileMeta[]>((acc, val) => {
  acc.push(...val)

  return acc
}, [])

export const readFsThreeList = (pathList: string[]) => Promise
  .all(pathList.map((url) => readFsThree(url)))
  .then(mergeFileLists)
