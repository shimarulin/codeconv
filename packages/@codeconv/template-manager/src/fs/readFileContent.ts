import { readFile } from 'fs/promises'
import { parsedPathToAbsolutePath } from '../path/parsedPathToAbsolutePath.js'
import type { FileMeta } from './readFsThree.js'

export interface FileDescription extends FileMeta {
  /**
   * The file content
   */
  content: string;
}

export const readFileContent = async (parsedPath: FileMeta): Promise<FileDescription> => {
  const content = await readFile(parsedPathToAbsolutePath(parsedPath), 'utf8')
  return {
    ...parsedPath,
    content,
  }
}
