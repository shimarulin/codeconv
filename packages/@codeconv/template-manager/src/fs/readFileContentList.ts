import { FileMeta } from './readFsThree.js'
import { readFileContent } from './readFileContent.js'

export const readFileContentList = (parsedPathList: FileMeta[]) => Promise
  .all(parsedPathList.map((parsedPath) => readFileContent(parsedPath)))
