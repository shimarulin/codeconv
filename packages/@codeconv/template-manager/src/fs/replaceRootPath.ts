import { FileDescription } from './readFileContent'

export const replaceRootPath = (targetPath: string, fileList: FileDescription[]) => {
  return fileList.map((file) => ({
    ...file,
    root: targetPath,
  }))
}
