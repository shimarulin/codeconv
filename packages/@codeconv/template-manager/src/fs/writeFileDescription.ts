import { mkdir, writeFile } from 'node:fs/promises'
import { join } from 'node:path'
import { type FileDescription } from './readFileContent.js'

export const writeFileDescription = async (file: FileDescription) => {
  const dirPath = join(file.root, file.dir)
  const filePath = join(dirPath, file.base)

  await mkdir(dirPath, {
    recursive: true,
  })
  await writeFile(filePath, file.content)

  return file
}
