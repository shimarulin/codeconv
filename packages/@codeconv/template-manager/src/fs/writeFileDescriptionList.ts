import { writeFileDescription } from './writeFileDescription.js'
import { type FileDescription } from './readFileContent.js'

export const writeFileDescriptionList = (fileList: FileDescription[]) => {
  return Promise.all(fileList.map((file) => writeFileDescription(file)))
}
