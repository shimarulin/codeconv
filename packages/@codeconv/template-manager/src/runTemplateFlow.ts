import { andThen, curry, pipeWith } from 'ramda'
import { readFsThreeList } from './fs/readFsThreeList.js'
import { readFileContentList } from './fs/readFileContentList.js'
import { replaceRootPath } from './fs/replaceRootPath'
import { writeFileDescriptionList } from './fs/writeFileDescriptionList'
import { type TemplateData, type TemplateHandler } from './handler/handleFileDescription.js'
import { handleFileDescriptionList } from './handler/handleFileDescriptionList.js'
import { getPathListFromStringOrURLList } from './path/getPathListFromStringOrURLList'

export const runTemplateFlow = (inputDirList: (string | URL)[], handlerList: TemplateHandler[], data: TemplateData, outputDir: string) => {
  const input = getPathListFromStringOrURLList(inputDirList)
  const applyHandlers = curry(handleFileDescriptionList)(handlerList, data)
  const setOutputDir = curry(replaceRootPath)(outputDir)

  return pipeWith(andThen, [
    readFsThreeList,
    readFileContentList,
    applyHandlers,
    setOutputDir,
    writeFileDescriptionList,
  ])(input)
}
