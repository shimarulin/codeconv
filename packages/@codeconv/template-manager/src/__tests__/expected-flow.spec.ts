import { rm } from 'fs/promises'
import { tmpdir } from 'os'
import { join } from 'path'
import { runTemplateFlow } from '../runTemplateFlow'

const pathToFlatTemplate = join(__dirname, '__fixtures__/flat-template')
const pathToDeepTemplate = join(__dirname, '__fixtures__/deep-template')
const pathToDeepTemplateDup = join(__dirname, '__fixtures__/deep-template-duplicate')
const pathToNotExistTemplate = join(__dirname, '__fixtures__/not-exist')

describe('Run template flow', () => {
  test('merge templates', async () => {
    const targetDir = join(tmpdir(), 'tpl-mt')
    const fileList = await runTemplateFlow(
      [
        pathToFlatTemplate,
        pathToDeepTemplate,
      ],
      [],
      {},
      targetDir,
    )

    fileList.forEach((fileMeta) => {
      expect(fileMeta).toMatchSnapshot()
    })

    await rm(targetDir, {
      recursive: true,
      force: true,
    })
  })

  test('merge duplicate templates', async () => {
    const targetDir = join(tmpdir(), 'tpl-mdt')
    const runner = async () => {
      return runTemplateFlow(
        [
          pathToFlatTemplate,
          pathToDeepTemplate,
          pathToDeepTemplateDup,
        ],
        [],
        {},
        targetDir,
      )
    }

    await expect(runner).rejects.toThrow(`File 'src/func/log.js' from template directory ${pathToDeepTemplateDup} already exist in ${pathToDeepTemplate}. You can't write more than one file to the same target path`)
  })

  test('handle non exist template', async () => {
    const targetDir = join(tmpdir(), 'tpl-mdt')
    const runner = async () => {
      return runTemplateFlow(
        [
          pathToFlatTemplate,
          pathToDeepTemplate,
          pathToNotExistTemplate,
        ],
        [],
        {},
        targetDir,
      )
    }

    await expect(runner).rejects.toThrow(`ENOENT: no such file or directory, opendir '${pathToNotExistTemplate}'`)
  })
})
