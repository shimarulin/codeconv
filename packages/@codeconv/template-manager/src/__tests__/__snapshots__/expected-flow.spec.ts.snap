// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run template flow merge templates 1`] = `
Object {
  "base": "LICENSE",
  "content": "The MIT License (MIT)

Copyright (c) 2022 Vyacheslav Shimarulin

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the \\"Software\\"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED \\"AS IS\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
",
  "dir": "./",
  "ext": "",
  "name": "LICENSE",
  "root": "/tmp/tpl-mt",
}
`;

exports[`Run template flow merge templates 2`] = `
Object {
  "base": "README.md",
  "content": "# Template manager test assets
",
  "dir": "./",
  "ext": ".md",
  "name": "README",
  "root": "/tmp/tpl-mt",
}
`;

exports[`Run template flow merge templates 3`] = `
Object {
  "base": "index.js",
  "content": "export {
  log,
} from './func/log.js'
",
  "dir": "./src",
  "ext": ".js",
  "name": "index",
  "root": "/tmp/tpl-mt",
}
`;

exports[`Run template flow merge templates 4`] = `
Object {
  "base": "log.js",
  "content": "export const log = (msg) => {
  console.log(msg)
}
",
  "dir": "./src/func",
  "ext": ".js",
  "name": "log",
  "root": "/tmp/tpl-mt",
}
`;

exports[`Run template flow merge templates 5`] = `
Object {
  "base": "package.json.json",
  "content": "{
  \\"name\\": \\"template-manager-test-assets\\",
  \\"private\\": true
}
",
  "dir": "./",
  "ext": ".json",
  "name": "package.json",
  "root": "/tmp/tpl-mt",
}
`;
